pkg/domain/infra/abi/quadlet.go

```diff
@@ -0,0 +1,764 @@
+//
+// Parameters:
+//   - ctx: Context for cancellation and timeout handling
+//   - quadlets: List of Quadlet names to remove, or empty slice if options.All is true
+//   - options: Configuration options including Force, Ignore, and All flags
+//
+// Returns:
+//   - QuadletRemoveReport: Contains lists of successfully removed quadlets and any errors encountered
+//   - error: Non-nil if a fatal error occurred that prevented the operation from completing
+//
+// The function supports:
+//   - Force removal of running Quadlet services
+//   - Ignoring missing Quadlet files (with Ignore option)
+//   - Removing all installed Quadlets (with All option)
+//   - Handling Quadlet applications (.app files containing multiple related Quadlets)
+func (ic *ContainerEngine) QuadletRemove(ctx context.Context, quadlets []string, options entities.QuadletRemoveOptions) (*entities.QuadletRemoveReport, error) {
+	report := entities.QuadletRemoveReport{
+		Errors:  make(map[string]error),
+		Removed: []string{},
+	}
+	removeList := []string{}
+	reverseMap, appMap, err := buildAppMap(systemdquadlet.GetInstallUnitDirPath(rootless.IsRootless()))
+	if err != nil {
+		return nil, fmt.Errorf("unable to build app map: %w", err)
+	}
+	quadletsRe := []string{}
+	// Process all `.app` files in arguments, if `.app` file
+	// is found then expand it to its respective quadlet files
+	// and remove it from the processing list.
+	for _, quadlet := range quadlets {
+		// Most likely this is an app
+		if strings.HasPrefix(quadlet, ".") && strings.HasSuffix(quadlet, ".app") {
+			files, ok := appMap[quadlet]
+			// Add all files of this application in to-be removed list.
+			if ok {
+				for _, file := range files {
+					if !systemdquadlet.IsExtSupported(file) {
+						removeList = append(removeList, file)
+					} else {
+						quadletsRe = append(quadletsRe, file)
+					}
+				}
+			}
+			// also add .app file itself to the remove list so it can
+			// be cleaned after removal of all components in the list
+			if !slices.Contains(removeList, quadlet) {
+				removeList = append(removeList, quadlet)
+			}
+		} else {
+			quadletsRe = append(quadletsRe, quadlet)
+		}
+	}
+	quadlets = quadletsRe
+	allQuadletPaths := make([]string, 0, len(quadlets))
+	allServiceNames := make([]string, 0, len(quadlets))
+	runningQuadlets := make([]string, 0, len(quadlets))
+	serviceNameToQuadletName := make(map[string]string)
+	needReload := false
+
+	// Early escape: if 0 quadlets are requested, bail immediately without error.
+	if len(quadlets) == 0 && !options.All {
+		return nil, errors.New("must provide at least 1 quadlet to remove")
+	}
+
+	// Is systemd available to the current user?
+	// We cannot proceed if not.
+	conn, err := systemd.ConnectToDBUS()
+	if err != nil {
+		return nil, fmt.Errorf("connecting to systemd dbus: %w", err)
+	}
+
+	if options.All {
+		allQuadlets := getAllQuadletPaths()
+		quadlets = allQuadlets
+	}
+
+	for i := 0; i < len(quadlets); i++ {
```

Comment:
why is this using an index loop, we should be using for range in go for a simple string slice iteration.

Reply:
It is extremely hard to reason about this entire loop block. With something like this there is the obvious question of how we ensure we do not loop forever if we keep adding entries.
Yes I cannot see that this is broken but it does take quite some work for to actual parse and understand that block there.
Maybe this can be written in a more readable way that does not require such in loop modifications

Reply:
Can i refactor this later ? In a followup PR.